File: packages/ui/src/types/engagement.ts
/**
 * Engagement Types
 *
 * Types for engagement management features.
 */

import type {
  EngagementStatus,
  EngagementPriority,
  DataSourceReference,
} from '@/lib/api';

// Engagement entity
export interface Engagement {
  id: string;
  name: string;
  description?: string;
  objective: string;
  status: EngagementStatus;
  priority: EngagementPriority;
  data_sources: DataSourceReference[];
  agent_config?: EngagementAgentConfig;
  tags: string[];
  metadata?: Record<string, unknown>;
  requires_approval: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
  started_at?: string;
  completed_at?: string;
  error_message?: string;
}

// Engagement summary for list views
export interface EngagementSummary {
  id: string;
  name: string;
  status: EngagementStatus;
  priority: EngagementPriority;
  created_by: string;
  created_at: string;
  updated_at: string;
  task_count?: number;
  completion_percentage?: number;
}

// Agent configuration for engagement
export interface EngagementAgentConfig {
  model?: string;
  temperature?: number;
  max_tokens?: number;
  tools?: string[];
  system_prompt?: string;
  max_iterations?: number;
  timeout_seconds?: number;
}

// Form data for creating/editing engagements
export interface EngagementFormData {
  name: string;
  description?: string;
  objective: string;
  priority: EngagementPriority;
  data_sources: DataSourceReference[];
  agent_config?: EngagementAgentConfig;
  tags: string[];
  metadata?: Record<string, unknown>;
  requires_approval: boolean;
}

// Status transition rules
export const ENGAGEMENT_STATUS_TRANSITIONS: Record<EngagementStatus, EngagementStatus[]> = {
  draft: ['pending_approval', 'cancelled'],
  pending_approval: ['approved', 'cancelled'],
  approved: ['in_progress', 'cancelled'],
  in_progress: ['paused', 'completed', 'failed', 'cancelled'],
  paused: ['in_progress', 'cancelled'],
  completed: [],
  cancelled: [],
  failed: ['draft'],
};

// Status display configuration
export const ENGAGEMENT_STATUS_CONFIG: Record<
  EngagementStatus,
  { label: string; color: string; bgColor: string }
> = {
  draft: { label: 'Draft', color: 'text-gray-600', bgColor: 'bg-gray-100' },
  pending_approval: { label: 'Pending Approval', color: 'text-yellow-600', bgColor: 'bg-yellow-100' },
  approved: { label: 'Approved', color: 'text-blue-600', bgColor: 'bg-blue-100' },
  in_progress: { label: 'In Progress', color: 'text-indigo-600', bgColor: 'bg-indigo-100' },
  paused: { label: 'Paused', color: 'text-orange-600', bgColor: 'bg-orange-100' },
  completed: { label: 'Completed', color: 'text-green-600', bgColor: 'bg-green-100' },
  cancelled: { label: 'Cancelled', color: 'text-gray-500', bgColor: 'bg-gray-100' },
  failed: { label: 'Failed', color: 'text-red-600', bgColor: 'bg-red-100' },
};

// Priority display configuration
export const ENGAGEMENT_PRIORITY_CONFIG: Record<
  EngagementPriority,
  { label: string; color: string; bgColor: string }
> = {
  low: { label: 'Low', color: 'text-gray-600', bgColor: 'bg-gray-100' },
  medium: { label: 'Medium', color: 'text-blue-600', bgColor: 'bg-blue-100' },
  high: { label: 'High', color: 'text-orange-600', bgColor: 'bg-orange-100' },
  critical: { label: 'Critical', color: 'text-red-600', bgColor: 'bg-red-100' },
};

// Helper type for engagement filters
export interface EngagementFilters {
  status?: EngagementStatus[];
  priority?: EngagementPriority[];
  tags?: string[];
  created_by?: string;
  date_range?: {
    start: string;
    end: string;
  };
}

File: packages/api/src/api/schemas/engagement.py
"""
Engagement-related Pydantic schemas.
"""
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum


class EngagementStatus(str, Enum):
    """Engagement lifecycle status."""
    DRAFT = "draft"
    PENDING_APPROVAL = "pending_approval"
    APPROVED = "approved"
    IN_PROGRESS = "in_progress"
    PAUSED = "paused"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    FAILED = "failed"


class EngagementPriority(str, Enum):
    """Engagement priority levels."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class DataSourceReference(BaseModel):
    """Reference to a data source used in an engagement."""
    source_id: str = Field(description="Data source ID")
    source_type: str = Field(description="Type of data source")
    access_mode: str = Field(default="read", description="Access mode: read, write, or readwrite")
    config: Optional[Dict[str, Any]] = Field(default=None, description="Source-specific config")


class EngagementCreate(BaseModel):
    """Schema for creating a new engagement."""
    name: str = Field(min_length=1, max_length=255, description="Engagement name")
    description: Optional[str] = Field(default=None, max_length=2000, description="Description")
    objective: str = Field(min_length=1, description="Main objective of the engagement")
    priority: EngagementPriority = Field(default=EngagementPriority.MEDIUM)
    data_sources: List[DataSourceReference] = Field(
        default_factory=list,
        description="Data sources to use"
    )
    agent_config: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Configuration for the AI agent"
    )
    tags: List[str] = Field(default_factory=list, description="Tags for categorization")
    metadata: Optional[Dict[str, Any]] = Field(default=None, description="Additional metadata")
    requires_approval: bool = Field(
        default=True,
        description="Whether this engagement requires approval before execution"
    )


class EngagementUpdate(BaseModel):
    """Schema for updating an existing engagement."""
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = Field(default=None, max_length=2000)
    objective: Optional[str] = Field(default=None, min_length=1)
    priority: Optional[EngagementPriority] = None
    data_sources: Optional[List[DataSourceReference]] = None
    agent_config: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


class EngagementResponse(BaseModel):
    """Schema for engagement response."""
    id: str = Field(description="Unique engagement ID")
    name: str
    description: Optional[str] = None
    objective: str
    status: EngagementStatus
    priority: EngagementPriority
    data_sources: List[DataSourceReference] = Field(default_factory=list)
    agent_config: Optional[Dict[str, Any]] = None
    tags: List[str] = Field(default_factory=list)
    metadata: Optional[Dict[str, Any]] = None
    requires_approval: bool
    created_by: str = Field(description="User ID of creator")
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None

    model_config = {"from_attributes": True}


class EngagementSummary(BaseModel):
    """Lightweight engagement summary for list views."""
    id: str
    name: str
    status: EngagementStatus
    priority: EngagementPriority
    created_by: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EngagementStatusUpdate(BaseModel):
    """Schema for updating engagement status."""
    status: EngagementStatus = Field(description="New status")
    reason: Optional[str] = Field(default=None, description="Reason for status change")


class EngagementExecutionResult(BaseModel):
    """Result of an engagement execution."""
    engagement_id: str
    status: EngagementStatus
    started_at: datetime
    completed_at: Optional[datetime] = None
    outputs: Optional[Dict[str, Any]] = None
    artifacts: List[str] = Field(default_factory=list, description="Generated artifact IDs")
    error_message: Optional[str] = None
    execution_time_seconds: Optional[float] = None
